### ISSUE TITLE
Question

### ISSUE THREAD

**Issue by extragoya** [raun1/Complementary_Segmentation_Network#1]
Thanks for sharing on github! This is not an issue, just a question. 

It seems that the complementary mask and the reconstruction output are at cross purposes, right?  Because, the former, on its own, would tend to produce an output that, when combined with the normal mask, would sum to a white image. This would obviously be uninformative to the reconstruction branch. 

So the complementary mask ends up being a bit of a compromise between containing enough information for the reconstruction while also trying to satisfy the negative Dice. Do I understand this correctly? 

I wonder, did you ever try training without the complementary mask loss? That is, just having the standard ground Dice loss and the reconstruction loss? So the complementary output is free to add whatever information it needs for reconstruction purposes? It may end up being "complementary" enough to be robust to pathologies. It would also help test whether your performance is due to (a) the network creating features for outside-brain regions; (b) the network using reconstruction to help regularize any mask output based on the pixel space of the input image; (c) or both.

**Comment by raun1** [comment]
Hello ! 

So the complementary branch has positive dice or no overlap with the ground truth mask( I guess the names are confusing , so normal Dice for neural networks are implemented as a negative dice and it gives overlap with gt mask and simply returns the negative value  of it for neural networks can only minimize. So I called that function which does positive dice as negative dice explicitly since it's simply a negative of the normal Dice but for evaluation purpose it's a value >=0). Now Since positive dice is not informative of what it should actually try to learn (since optimal value of it is 0 since it's not defined for negative values) we kind of need the reconstruction branch to make it learn something useful. 

Thank you

**Comment by raun1** [comment]
As for trying to use simply a segmentation branch and reconstruction branch and skipping the whole complementary branch, we didn't try it but it was always something I thought to try out honestly :)

**Comment by raun1** [comment]
As for the second case. It is interesting if we don't specify anything as a complementary loss and something I will check it out during winter break and email you (semester is kinda busy sadly). I am curious too what it will give maybe an encoding of the whole image ? Which can be useful too.

**Comment by extragoya** [comment]
Ah, yes I see. The complementary loss is an underdetermined problem, there are many masks (they don't even have to be masks in fact), that satisfy zero overlap, which would minimize the complementary loss. Basically, the complementary mask has to satisfy two conditions at once (a) no overlap and (b) provide additional information to aid reconstruction. 

Yes, I wonder if it's actually the reconstruction, and not the complementary part, that is somehow regularizing your brain segmentation output and allowing it to be robust to pathologies. Hard to say without testing it. 

Now that I think about it though, if you don't specify anything as a complementary loss, then it's possible that your network will just learn an identify function, making the CO the same as the image, which would then defeat the purpose (and make it an easy task for your reconstruction network!). 

In any event, thank you for getting back to me!
